### PR TITLE
use hpp for c++ code

### PR DESCRIPTION
**Pull request by wjwwood**
For consistency and to not confuse `cppcheck`.

Connects to ros2/ros2#368

### CODE REVIEW

**Comment by wjwwood** [comment]
CI:

- Linux:
  - [Build Status](http://ci.ros2.org/job/ci_linux/2788/)
- Linux-aarch64:
  - [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/314/)
- macOS:
  - [Build Status](http://ci.ros2.org/job/ci_osx/2260/)
- Windows:
  - [Build Status](http://ci.ros2.org/job/ci_windows/2926/)

The other changes (to force cppcheck to interpret `.h` as C++) will need to be done separately once the newly updated Windows machines are turned on (it won't fail until that happens so it will be hard to test). Anyways these changes can be done separately.

**Comment by mikaelarguedas** [comment]
Do you think it's worth it to split the renaming of the files and the change of the include rules into different commits so that the affected files preserve their history? (I know that means that the state of the code in between the 2 commits won't compile but I think it's worth it)

**Comment by allenh1** [comment]
@mikaelguerdas the diff makes it look like it picked up on the renaming. I think it's safe.

**Comment by mikaelarguedas** [comment]
Id did, unfortunately the history is still this single commit : https://github.com/ros2/rmw_fastrtps/commits/2403a757da9e55e8cc31979909cbb6abba4452a0/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/MessageTypeSupport.hpp So my concern remains

**Comment by allenh1** [comment]
Ah, I see. Ya, that would be rather unfortunate.

**Comment by wjwwood** [comment]
@mikaelarguedas are you sure that having a separate commit would change anything? I'm pretty sure that GitHub would have to implement the `--follow` option of `git log`:

https://stackoverflow.com/questions/5646174/how-to-make-github-follow-directory-history-after-renames

I think you can follow the file rename through the git blame interface on GitHub:

https://github.com/ros2/rmw_fastrtps/blame/2403a757da9e55e8cc31979909cbb6abba4452a0/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/MessageTypeSupport.hpp

**Comment by mikaelarguedas** [comment]
Oh I thought id did but I stand corrected by Github UI o_O (https://github.com/ros2/joystick_drivers/commits/ros2/joy/src/joy_node_linux.cpp). Totally fine as is then, sorry for the noise.

**Comment by wjwwood** [comment]
Forgot to run `ament_cpplint`, doh. Fixing up now, I'll have to run CI again.

**Comment by wjwwood** [comment]
New CI:

- Linux:
  - [Build Status](http://ci.ros2.org/job/ci_linux/2789/)
- Linux-aarch64:
  - [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/315/)
- macOS:
  - [Build Status](http://ci.ros2.org/job/ci_osx/2261/)
- Windows:
  - [Build Status](http://ci.ros2.org/job/ci_windows/2927/)

@mikaelarguedas (or anyone else) can you re-review this now?

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 1]
```diff
@@ -344,8 +344,8 @@ size_t get_array_size_and_assign_field(
   size_t space)
 {
   std::vector<unsigned char> * vector = reinterpret_cast<std::vector<unsigned char> *>(field);
-  void * ptr = (void *)sub_members_size;
-  size_t vsize = vector->size() / (size_t)align_(max_align, 0, ptr, space);
+  void * ptr = reinterpret_cast<void *>(sub_members_size);
```
Comment (wjwwood) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: btw, we're casting a `size_t` to a `void *` here :nauseated_face:.

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 2]
```diff
@@ -376,7 +377,7 @@ bool TypeSupport<MembersType>::serializeROSmessage(
 
   for (uint32_t i = 0; i < members->member_count_; ++i) {
     const auto member = members->members_ + i;
-    void * field = (char *)ros_message + member->offset_;
+    void * field = const_cast<char *>(static_cast<const char *>(ros_message)) + member->offset_;
```
Comment (wjwwood) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: Also, I could not convert the `const void *` directly to `char *` so I did it in two steps...

**Comment by wjwwood** [comment]
Can someone please re-review this one?

**Comment by wjwwood** [comment]
I'm gonna merge this one, since it is green without the related cppcheck changes in the other pr's. Note that linux-aarch is unstable because I had to restart it but only after I pushed some, but not all, of the other branches. At any rate I'm sure this pr is good to go.

**Pull request merged**